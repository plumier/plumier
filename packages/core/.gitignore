!src/generic-controller/filter-grammar.js