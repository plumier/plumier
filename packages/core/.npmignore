test
src
.npmignore
.gitignore
tsconfig.json
package-lock.json
tsconfig.build.json
yarn-error.log
yarn.lock