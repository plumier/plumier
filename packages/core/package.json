{
  "name": "@plumier/core",
  "version": "1.0.0-beta.9",
  "description": "Delightful Node.js Rest Framework",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "MVC",
    "Koa",
    "IoC",
    "Dependency Injection",
    "Web API",
    "JSON Service",
    "TDD"
  ],
  "scripts": {
    "compile": "tsc -p tsconfig.build.json"
  },
  "author": "Ketut Sandiarsa",
  "license": "MIT",
  "dependencies": {
    "@types/glob": "^7.1.1",
    "@types/koa": "^2.0.50",
    "chalk": "^2.4.2",
    "glob": "^7.1.4",
    "path-to-regexp": "^3.1.0",
    "tinspector": "^2.2.7",
    "tslib": "^1.10.0",
    "typedconverter": "^1.0.0"
  },
  "bugs": {
    "url": "https://github.com/plumier/plumier/issues"
  },
  "repository": {
    "url": "https://github.com/plumier/plumier"
  },
  "homepage": "https://plumierjs.com",
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "9d69b5621b7e2e777818e5e36c13ecba0c8d8204",
  "devDependencies": {
    "upath": "^1.2.0"
  }
}
