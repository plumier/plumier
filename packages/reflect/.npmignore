.vscode
coverage
scripts
src 
test
.gitignore
.npmignore
.travis.yml
package-lock.json
tsconfig.build.json
tsconfig.json
yarn-error.log
yarn.lock