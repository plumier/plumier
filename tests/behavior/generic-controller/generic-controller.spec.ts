import { GenericController, Class, GenericOneToManyController, crud, createRoutesFromEntities, OneToManyDecorator, RouteInfo, route } from "@plumier/core"
import reflect, { generic, metadata, decorateClass } from "tinspector"
import { transform } from "@plumier/swagger"

describe("Generic Controller", () => {
    class User {
        @reflect.noop()
        name: string

        @reflect.noop()
        email: string
    }
    it("Should able to use Number id", () => {
        @generic.type(User, Number)
        class UsersController extends GenericController<User, Number> { }
        const meta = reflect(UsersController)
        expect(metadata.getMethods(meta)).toMatchSnapshot()
        const idResult = reflect(meta.methods.find(x => x.name === "save")!.returnType as Class)
        expect(metadata.getProperties(idResult)).toMatchSnapshot()
    })
    it("Should able to use String id", () => {
        @generic.type(User, String)
        class UsersController extends GenericController<User, String> { }
        const meta = reflect(UsersController)
        expect(metadata.getMethods(meta)).toMatchSnapshot()
        const idResult = reflect(meta.methods.find(x => x.name === "save")!.returnType as Class)
        expect(metadata.getProperties(idResult)).toMatchSnapshot()
    })
    it("Should invocable", () => {
        @generic.type(User, Number)
        class TypeOrmGenericController extends GenericController<User, Number>{
            list(offset: number = 0, limit: number = 50, query: User) {
                return super.list(offset, limit, query)
            }
            async save(data: User) {
                return super.save(data)
            }
            get(id: number) {
                return super.get(id)
            }
            async modify(id: number, data: User) {
                return super.modify(id, data)
            }
            async delete(id: number) {
                return super.delete(id)
            }
        }
        const ctl = new TypeOrmGenericController()
        ctl.list(1, 10, {} as User)
        ctl.list(undefined, undefined, {} as User)
        ctl.save({} as User)
        ctl.get(1)
        ctl.modify(1, {} as User)
        ctl.delete(1)
    })
})

describe("Generic One To Many Controller", () => {
    class User {
        @reflect.noop()
        name: string

        @reflect.noop()
        email: string

        @reflect.type(x => [Animal])
        animals: Animal[]
    }
    class Animal {
        @reflect.noop()
        name: string
    }
    it("Should able to use Number id", () => {
        @generic.type(User, Animal, Number, Number)
        class UsersController extends GenericOneToManyController<User, Animal, Number, Number> { }
        const meta = reflect(UsersController)
        expect(metadata.getMethods(meta)).toMatchSnapshot()
        const idResult = reflect(meta.methods.find(x => x.name === "save")!.returnType as Class)
        expect(metadata.getProperties(idResult)).toMatchSnapshot()
    })
    it("Should able to use String id", () => {
        @generic.type(User, Animal, Number, String)
        class UsersController extends GenericOneToManyController<User, Animal, Number, String> { }
        const meta = reflect(UsersController)
        expect(metadata.getMethods(meta)).toMatchSnapshot()
        const idResult = reflect(meta.methods.find(x => x.name === "save")!.returnType as Class)
        expect(metadata.getProperties(idResult)).toMatchSnapshot()
    })
    it("Should able to use Number pid", () => {
        @generic.type(User, Animal, Number, Number)
        class UsersController extends GenericOneToManyController<User, Animal, Number, Number> { }
        const meta = reflect(UsersController)
        expect(metadata.getMethods(meta)).toMatchSnapshot()
        const idResult = reflect(meta.methods.find(x => x.name === "save")!.returnType as Class)
        expect(metadata.getProperties(idResult)).toMatchSnapshot()
    })
    it("Should able to use String pid", () => {
        @generic.type(User, Animal, String, Number)
        class UsersController extends GenericOneToManyController<User, Animal, String, Number> { }
        const meta = reflect(UsersController)
        expect(metadata.getMethods(meta)).toMatchSnapshot()
        const idResult = reflect(meta.methods.find(x => x.name === "save")!.returnType as Class)
        expect(metadata.getProperties(idResult)).toMatchSnapshot()
    })
    it("Should invocable", () => {
        @generic.type(User, Animal, Number, Number)
        @decorateClass(<OneToManyDecorator>{ propertyName: "animals", parentType: User, kind: "GenericDecoratorOneToMany", type: x => UsersController })
        class UsersController extends GenericOneToManyController<User, Animal, Number, Number> {
            list(pid: number, offset: number = 0, limit: number = 50, query: User) {
                return super.list(pid, offset, limit, query)
            }
            async save(pid: number, data: Animal) {
                return super.save(pid, data)
            }
            get(pid: number, id: number) {
                return super.get(pid, id)
            }
            async modify(pid: number, id: number, data: Animal) {
                return super.modify(pid, id, data)
            }
            async delete(pid: number, id: number) {
                return super.delete(pid, id)
            }
        }
        const ctl = new UsersController()
        ctl.list(1, 1, 10, {} as User)
        ctl.list(1, undefined, undefined, {} as User)
        ctl.save(1, {} as User)
        ctl.get(1, 1)
        ctl.modify(1, 1, {} as User)
        ctl.delete(1, 1)
    })
    it("Should throw error when no OneToManyDecorator provided", () => {
        @generic.type(User, Animal, Number, Number)
        class UsersController extends GenericOneToManyController<User, Animal, Number, Number> {
            list(pid: number, offset: number = 0, limit: number = 50, query: User) {
                return super.list(pid, offset, limit, query)
            }
            async save(pid: number, data: Animal) {
                return super.save(pid, data)
            }
            get(pid: number, id: number) {
                return super.get(pid, id)
            }
            async modify(pid: number, id: number, data: Animal) {
                return super.modify(pid, id, data)
            }
            async delete(pid: number, id: number) {
                return super.delete(pid, id)
            }
        }
        expect(() => new UsersController()).toThrowErrorMatchingSnapshot()
    })
})

describe("Rote Generator", () => {
    class Animal {
        @reflect.noop()
        name: string
    }
    class User {
        @reflect.noop()
        name: string

        @reflect.noop()
        email: string

        @crud.oneToMany(x => Animal)
        @reflect.type(x => [Animal])
        animals: Animal[]
    }
    it("Should generate routes properly", () => {
        @generic.template("T", "TID")
        @generic.type("T", "TID")
        class UsersController<T, TID> extends GenericController<T, TID> { }
        @generic.template("P", "T", "PID", "TID")
        @generic.type("P", "T", "PID", "TID")
        class UsersAnimalsController<P, T, PID, TID> extends GenericOneToManyController<P, T, PID, TID> { }
        const routes = createRoutesFromEntities([User, Animal], UsersController, UsersAnimalsController, x => x)
        expect(routes.map(x => ({ method: x.method, path: x.url }))).toMatchSnapshot()
    })
})

describe("Swagger", () => {
    describe("Generic Controller", () => {
        it("Should generate GET /animal properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            const routes = createRoutesFromEntities([Animal], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/animal"].get.parameters).toMatchSnapshot()
            expect(spec.paths["/animal"].get.tags).toMatchSnapshot()
        })
        it("Should generate POST /animal properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            const routes = createRoutesFromEntities([Animal], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/animal"].post.requestBody).toMatchSnapshot()
            expect(spec.paths["/animal"].post.tags).toMatchSnapshot()
        })
        it("Should generate GET /animal/:id properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            const routes = createRoutesFromEntities([Animal], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/animal/{id}"].get.parameters).toMatchSnapshot()
            expect(spec.paths["/animal/{id}"].get.tags).toMatchSnapshot()
        })
        it("Should generate DELETE /animal/:id properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            const routes = createRoutesFromEntities([Animal], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/animal/{id}"].delete.parameters).toMatchSnapshot()
            expect(spec.paths["/animal/{id}"].delete.tags).toMatchSnapshot()
        })
        it("Should generate PUT /animal/:id properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            const routes = createRoutesFromEntities([Animal], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/animal/{id}"].put.parameters).toMatchSnapshot()
            expect(spec.paths["/animal/{id}"].put.tags).toMatchSnapshot()
        })
        it("Should generate PATCH /animal/:id properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            const routes = createRoutesFromEntities([Animal], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/animal/{id}"].put.parameters).toMatchSnapshot()
            expect(spec.paths["/animal/{id}"].put.tags).toMatchSnapshot()
        })
    })
    describe("One To Many Generic Controller", () => {
        it("Should generate GET /animal properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            class Client {
                @reflect.noop()
                name:string
                @crud.oneToMany(x => Animal)
                @reflect.type(x => [Animal])
                animals: Animal[]
            }
            const routes = createRoutesFromEntities([Animal, Client], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/client/{pid}/animals"].get.parameters).toMatchSnapshot()
            expect(spec.paths["/client/{pid}/animals"].get.tags).toMatchSnapshot()
        })
        it("Should generate POST /animal properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            class Client {
                @reflect.noop()
                name:string
                @crud.oneToMany(x => Animal)
                @reflect.type(x => [Animal])
                animals: Animal[]
            }
            const routes = createRoutesFromEntities([Animal, Client], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/client/{pid}/animals"].post.requestBody).toMatchSnapshot()
            expect(spec.paths["/client/{pid}/animals"].post.tags).toMatchSnapshot()
        })
        it("Should generate GET /animal/:id properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            class Client {
                @reflect.noop()
                name:string
                @crud.oneToMany(x => Animal)
                @reflect.type(x => [Animal])
                animals: Animal[]
            }
            const routes = createRoutesFromEntities([Animal, Client], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/client/{pid}/animals/{id}"].get.parameters).toMatchSnapshot()
            expect(spec.paths["/client/{pid}/animals/{id}"].get.tags).toMatchSnapshot()
        })
        it("Should generate DELETE /animal/:id properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            class Client {
                @reflect.noop()
                name:string
                @crud.oneToMany(x => Animal)
                @reflect.type(x => [Animal])
                animals: Animal[]
            }
            const routes = createRoutesFromEntities([Animal, Client], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/client/{pid}/animals/{id}"].delete.parameters).toMatchSnapshot()
            expect(spec.paths["/client/{pid}/animals/{id}"].delete.tags).toMatchSnapshot()
        })
        it("Should generate PUT /animal/:id properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            class Client {
                @reflect.noop()
                name:string
                @crud.oneToMany(x => Animal)
                @reflect.type(x => [Animal])
                animals: Animal[]
            }
            const routes = createRoutesFromEntities([Animal, Client], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/client/{pid}/animals/{id}"].put.parameters).toMatchSnapshot()
            expect(spec.paths["/client/{pid}/animals/{id}"].put.tags).toMatchSnapshot()
        })
        it("Should generate PATCH /animal/:id properly", () => {
            class Animal {
                @reflect.noop()
                name: string
            }
            class Client {
                @reflect.noop()
                name:string
                @crud.oneToMany(x => Animal)
                @reflect.type(x => [Animal])
                animals: Animal[]
            }
            const routes = createRoutesFromEntities([Animal, Client], GenericController, GenericOneToManyController, x => x)
            const spec = transform(routes, { map: new Map(), config: {} as any })
            expect(spec.paths["/client/{pid}/animals/{id}"].put.parameters).toMatchSnapshot()
            expect(spec.paths["/client/{pid}/animals/{id}"].put.tags).toMatchSnapshot()
        })
    })
})