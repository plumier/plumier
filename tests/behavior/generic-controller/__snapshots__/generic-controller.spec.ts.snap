// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic Controller Should able to use Number id 1`] = `
Array [
  Object {
    "name": "findByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "list",
    "pars": Array [
      Object {
        "name": "offset",
        "type": Number,
      },
      Object {
        "name": "limit",
        "type": Number,
      },
      Object {
        "name": "query",
        "type": User,
      },
    ],
    "type": Array [
      User,
    ],
  },
  Object {
    "name": "save",
    "pars": Array [
      Object {
        "name": "data",
        "type": User,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "get",
    "pars": Array [
      Object {
        "name": "id",
        "type": Number,
      },
    ],
    "type": User,
  },
  Object {
    "name": "modify",
    "pars": Array [
      Object {
        "name": "id",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": User,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "replace",
    "pars": Array [
      Object {
        "name": "id",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": User,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "delete",
    "pars": Array [
      Object {
        "name": "id",
        "type": Number,
      },
    ],
    "type": DynamicType,
  },
]
`;

exports[`Generic Controller Should able to use Number id 2`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
]
`;

exports[`Generic Controller Should able to use String id 1`] = `
Array [
  Object {
    "name": "findByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "list",
    "pars": Array [
      Object {
        "name": "offset",
        "type": Number,
      },
      Object {
        "name": "limit",
        "type": Number,
      },
      Object {
        "name": "query",
        "type": User,
      },
    ],
    "type": Array [
      User,
    ],
  },
  Object {
    "name": "save",
    "pars": Array [
      Object {
        "name": "data",
        "type": User,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "get",
    "pars": Array [
      Object {
        "name": "id",
        "type": String,
      },
    ],
    "type": User,
  },
  Object {
    "name": "modify",
    "pars": Array [
      Object {
        "name": "id",
        "type": String,
      },
      Object {
        "name": "data",
        "type": User,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "replace",
    "pars": Array [
      Object {
        "name": "id",
        "type": String,
      },
      Object {
        "name": "data",
        "type": User,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "delete",
    "pars": Array [
      Object {
        "name": "id",
        "type": String,
      },
    ],
    "type": DynamicType,
  },
]
`;

exports[`Generic Controller Should able to use String id 2`] = `
Array [
  Object {
    "name": "id",
    "type": String,
  },
]
`;

exports[`Generic One To Many Controller Should able to use Number id 1`] = `
Array [
  Object {
    "name": "findByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "findParentByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "list",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "offset",
        "type": Number,
      },
      Object {
        "name": "limit",
        "type": Number,
      },
      Object {
        "name": "query",
        "type": Animal,
      },
    ],
    "type": Array [
      Animal,
    ],
  },
  Object {
    "name": "save",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "get",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": Number,
      },
    ],
    "type": Animal,
  },
  Object {
    "name": "modify",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "replace",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "delete",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": Number,
      },
    ],
    "type": DynamicType,
  },
]
`;

exports[`Generic One To Many Controller Should able to use Number id 2`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
]
`;

exports[`Generic One To Many Controller Should able to use Number pid 1`] = `
Array [
  Object {
    "name": "findByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "findParentByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "list",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "offset",
        "type": Number,
      },
      Object {
        "name": "limit",
        "type": Number,
      },
      Object {
        "name": "query",
        "type": Animal,
      },
    ],
    "type": Array [
      Animal,
    ],
  },
  Object {
    "name": "save",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "get",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": Number,
      },
    ],
    "type": Animal,
  },
  Object {
    "name": "modify",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "replace",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "delete",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": Number,
      },
    ],
    "type": DynamicType,
  },
]
`;

exports[`Generic One To Many Controller Should able to use Number pid 2`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
]
`;

exports[`Generic One To Many Controller Should able to use String id 1`] = `
Array [
  Object {
    "name": "findByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "findParentByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "list",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "offset",
        "type": Number,
      },
      Object {
        "name": "limit",
        "type": Number,
      },
      Object {
        "name": "query",
        "type": Animal,
      },
    ],
    "type": Array [
      Animal,
    ],
  },
  Object {
    "name": "save",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "get",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": String,
      },
    ],
    "type": Animal,
  },
  Object {
    "name": "modify",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": String,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "replace",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": String,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "delete",
    "pars": Array [
      Object {
        "name": "pid",
        "type": Number,
      },
      Object {
        "name": "id",
        "type": String,
      },
    ],
    "type": DynamicType,
  },
]
`;

exports[`Generic One To Many Controller Should able to use String id 2`] = `
Array [
  Object {
    "name": "id",
    "type": String,
  },
]
`;

exports[`Generic One To Many Controller Should able to use String pid 1`] = `
Array [
  Object {
    "name": "findByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "findParentByIdOrNotFound",
    "pars": Array [
      Object {
        "name": "id",
        "type": Object,
      },
    ],
    "type": Promise,
  },
  Object {
    "name": "list",
    "pars": Array [
      Object {
        "name": "pid",
        "type": String,
      },
      Object {
        "name": "offset",
        "type": Number,
      },
      Object {
        "name": "limit",
        "type": Number,
      },
      Object {
        "name": "query",
        "type": Animal,
      },
    ],
    "type": Array [
      Animal,
    ],
  },
  Object {
    "name": "save",
    "pars": Array [
      Object {
        "name": "pid",
        "type": String,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "get",
    "pars": Array [
      Object {
        "name": "pid",
        "type": String,
      },
      Object {
        "name": "id",
        "type": Number,
      },
    ],
    "type": Animal,
  },
  Object {
    "name": "modify",
    "pars": Array [
      Object {
        "name": "pid",
        "type": String,
      },
      Object {
        "name": "id",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "replace",
    "pars": Array [
      Object {
        "name": "pid",
        "type": String,
      },
      Object {
        "name": "id",
        "type": Number,
      },
      Object {
        "name": "data",
        "type": Animal,
      },
    ],
    "type": DynamicType,
  },
  Object {
    "name": "delete",
    "pars": Array [
      Object {
        "name": "pid",
        "type": String,
      },
      Object {
        "name": "id",
        "type": Number,
      },
    ],
    "type": DynamicType,
  },
]
`;

exports[`Generic One To Many Controller Should able to use String pid 2`] = `
Array [
  Object {
    "name": "id",
    "type": Number,
  },
]
`;

exports[`Generic One To Many Controller Should throw error when no OneToManyDecorator provided 1`] = `"Configuration Error: UsersController doesn't decorated with OneToManyDecorator @decorateClass(<OneToManyDecorator>)"`;

exports[`Rote Generator Should generate routes properly 1`] = `
Array [
  Object {
    "method": "get",
    "path": "/user",
  },
  Object {
    "method": "post",
    "path": "/user",
  },
  Object {
    "method": "get",
    "path": "/user/:id",
  },
  Object {
    "method": "patch",
    "path": "/user/:id",
  },
  Object {
    "method": "put",
    "path": "/user/:id",
  },
  Object {
    "method": "delete",
    "path": "/user/:id",
  },
  Object {
    "method": "get",
    "path": "/user/:pid/animals",
  },
  Object {
    "method": "post",
    "path": "/user/:pid/animals",
  },
  Object {
    "method": "get",
    "path": "/user/:pid/animals/:id",
  },
  Object {
    "method": "patch",
    "path": "/user/:pid/animals/:id",
  },
  Object {
    "method": "put",
    "path": "/user/:pid/animals/:id",
  },
  Object {
    "method": "delete",
    "path": "/user/:pid/animals/:id",
  },
  Object {
    "method": "get",
    "path": "/animal",
  },
  Object {
    "method": "post",
    "path": "/animal",
  },
  Object {
    "method": "get",
    "path": "/animal/:id",
  },
  Object {
    "method": "patch",
    "path": "/animal/:id",
  },
  Object {
    "method": "put",
    "path": "/animal/:id",
  },
  Object {
    "method": "delete",
    "path": "/animal/:id",
  },
]
`;

exports[`Swagger Generic Controller Should generate DELETE /animal/:id properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "id",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`Swagger Generic Controller Should generate DELETE /animal/:id properly 2`] = `
Array [
  "Animal",
]
`;

exports[`Swagger Generic Controller Should generate GET /animal properly 1`] = `
Array [
  Object {
    "in": "query",
    "name": "name",
    "required": false,
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "description": undefined,
    "in": "query",
    "name": "offset",
    "required": false,
    "schema": Object {
      "type": "number",
    },
  },
  Object {
    "description": undefined,
    "in": "query",
    "name": "limit",
    "required": false,
    "schema": Object {
      "type": "number",
    },
  },
]
`;

exports[`Swagger Generic Controller Should generate GET /animal properly 2`] = `
Array [
  "Animal",
]
`;

exports[`Swagger Generic Controller Should generate GET /animal/:id properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "id",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`Swagger Generic Controller Should generate GET /animal/:id properly 2`] = `
Array [
  "Animal",
]
`;

exports[`Swagger Generic Controller Should generate PATCH /animal/:id properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "id",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`Swagger Generic Controller Should generate PATCH /animal/:id properly 2`] = `
Array [
  "Animal",
]
`;

exports[`Swagger Generic Controller Should generate POST /animal properly 1`] = `
Object {
  "content": Object {
    "application/json": Object {
      "schema": Object {
        "$ref": "#/components/schemas/Animal",
        "type": "object",
      },
    },
    "application/x-www-form-urlencoded": Object {
      "schema": Object {
        "$ref": "#/components/schemas/Animal",
        "type": "object",
      },
    },
  },
  "required": true,
}
`;

exports[`Swagger Generic Controller Should generate POST /animal properly 2`] = `
Array [
  "Animal",
]
`;

exports[`Swagger Generic Controller Should generate PUT /animal/:id properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "id",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`Swagger Generic Controller Should generate PUT /animal/:id properly 2`] = `
Array [
  "Animal",
]
`;

exports[`Swagger One To Many Generic Controller Should generate DELETE /animal/:id properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "pid",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "description": undefined,
    "in": "path",
    "name": "id",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`Swagger One To Many Generic Controller Should generate DELETE /animal/:id properly 2`] = `
Array [
  "Client",
]
`;

exports[`Swagger One To Many Generic Controller Should generate GET /animal properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "pid",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "in": "query",
    "name": "name",
    "required": false,
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "description": undefined,
    "in": "query",
    "name": "offset",
    "required": false,
    "schema": Object {
      "type": "number",
    },
  },
  Object {
    "description": undefined,
    "in": "query",
    "name": "limit",
    "required": false,
    "schema": Object {
      "type": "number",
    },
  },
]
`;

exports[`Swagger One To Many Generic Controller Should generate GET /animal properly 2`] = `
Array [
  "Client",
]
`;

exports[`Swagger One To Many Generic Controller Should generate GET /animal/:id properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "pid",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "description": undefined,
    "in": "path",
    "name": "id",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`Swagger One To Many Generic Controller Should generate GET /animal/:id properly 2`] = `
Array [
  "Client",
]
`;

exports[`Swagger One To Many Generic Controller Should generate PATCH /animal/:id properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "pid",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "description": undefined,
    "in": "path",
    "name": "id",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`Swagger One To Many Generic Controller Should generate PATCH /animal/:id properly 2`] = `
Array [
  "Client",
]
`;

exports[`Swagger One To Many Generic Controller Should generate POST /animal properly 1`] = `
Object {
  "content": Object {
    "application/json": Object {
      "schema": Object {
        "$ref": "#/components/schemas/Animal",
        "type": "object",
      },
    },
    "application/x-www-form-urlencoded": Object {
      "schema": Object {
        "$ref": "#/components/schemas/Animal",
        "type": "object",
      },
    },
  },
  "required": true,
}
`;

exports[`Swagger One To Many Generic Controller Should generate POST /animal properly 2`] = `
Array [
  "Client",
]
`;

exports[`Swagger One To Many Generic Controller Should generate PUT /animal/:id properly 1`] = `
Array [
  Object {
    "description": undefined,
    "in": "path",
    "name": "pid",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
  Object {
    "description": undefined,
    "in": "path",
    "name": "id",
    "required": true,
    "schema": Object {
      "type": "string",
    },
  },
]
`;

exports[`Swagger One To Many Generic Controller Should generate PUT /animal/:id properly 2`] = `
Array [
  "Client",
]
`;
